### PR TITLE
JAMES-2530 JMAP CRUD APIs for Filters

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Comment by chibenwa** [comment]
:green_apple: on new CI

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/GetFilterResponse.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            return this;
+        }
+
+        public Builder state(String state) {
+            this.state = state;
+            return this;
+        }
+
+        public Builder rules(List<Rule> rules) {
+            this.singleton.addAll(rules.stream()
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/GetFilterResponse.java]: The singleton is a list so... It's a bit the spec fault!

I'll try to cheat with the builder our code is using

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/GetMessageListRequest.java [hunk 2]
```diff
@@ -80,7 +80,7 @@ public Builder collapseThreads(boolean collapseThreads) {
         public Builder position(long position) {
             this.position = Optional.of(
                 Number.DEFAULT_FACTORY.from(position)
-                    .orElseThrow(() -> new IllegalArgumentException(Number.VALIDATION_MESSAGE)));
+                        .orElseThrow(() -> new IllegalArgumentException(Number.VALIDATION_MESSAGE)));
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/GetMessageListRequest.java]: unrelated

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetFilterMethod.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+            .build());
+    }
+
+    private void ensureNoMultiMailboxRules(ImmutableList<Rule> rules) throws MultipleMailboxIdException {
+        ImmutableList<Rule.Id> idWithMultipleMailboxes = rules.stream()
+            .filter(rule -> rule.getAction().getAppendInMailboxes().getMailboxIds().size() > 1)
+            .map(Rule::getId)
+            .collect(ImmutableList.toImmutableList());
+
+        if (!idWithMultipleMailboxes.isEmpty()) {
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetFilterMethod.java]: there is a mix of multiMailbox and MultipleMailboxes in this file, can you fix it please?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/FilterTest.java [hunk 4]
```diff
@@ -0,0 +1,575 @@
...
+            .post("/jmap");
+        with()
+            .header("Authorization", accessToken.serialize())
+            .body("[[" +
+                  "  \"setFilter\", " +
+                  "  {" +
+                  "    \"singleton\": [" +
+                  "    {" +
+                  "      \"id\": \"42-ac\"," +
+                  "      \"name\": \"My first rule\"," +
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/FilterTest.java]: first should be first and last should be last, no?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/FilterTest.java [hunk 5]
```diff
@@ -0,0 +1,575 @@
...
+                  "\"#0\"" +
+                  "]]")
+        .when()
+            .post("/jmap")
+        .then()
+            .statusCode(200)
+            .body(NAME, equalTo("filter"))
+            .body(ARGUMENTS + ".singleton", hasSize(1))
+            .body(ARGUMENTS + ".singleton[0].id", equalTo("42-ac"))
+            .body(ARGUMENTS + ".singleton[0].name", equalTo("My first rule"))
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/FilterTest.java]: and so here should be last

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/FilterTest.java [hunk 6]
```diff
@@ -0,0 +1,575 @@
...
+        .when()
+            .post("/jmap")
+        .then()
+            .body(NAME, equalTo("error"))
+            .body(ARGUMENTS + ".type", equalTo("invalidArguments"))
+            .body(ARGUMENTS + ".description", equalTo("The field 'ifInState' of 'SetFilterRequest' is not supported"));
+    }
+
+    @Test
+    public void getFilterFilterShouldRetrievePreviouslyStoredRules() {
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/FilterTest.java]: s/FilterFilter/Filter

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/FilterTest.java [hunk 7]
```diff
@@ -0,0 +1,575 @@
...
+            .body(ARGUMENTS + ".singleton[1].id", equalTo("3000-34e"))
+            .body(ARGUMENTS + ".singleton[1].name", equalTo("My last rule"))
+            .body(ARGUMENTS + ".singleton[1].condition.field", equalTo("subject"))
+            .body(ARGUMENTS + ".singleton[1].condition.comparator", equalTo("contains"))
+            .body(ARGUMENTS + ".singleton[1].condition.value", equalTo("question"))
+            .body(ARGUMENTS + ".singleton[1].action.appendIn.mailboxIds", containsInAnyOrder(m2.serialize()));
+    }
+
+    @Test
+    public void setFilterFilterShouldClearPreviouslyStoredRulesWhenEmptyBody() {
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/FilterTest.java]: idem

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetFilterMethod.java [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        filteringManagement.defineRulesForUser(user, rules);
+
+        return Stream.of(JmapResponse.builder()
+            .clientId(clientId)
+            .responseName(RESPONSE_NAME)
+            .response(SetFilterResponse.updated())
+            .build());
+    }
+
+    private void ensureNoMultipleMailboxRules(ImmutableList<Rule> rules) throws MultipleMailboxIdException {
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetFilterMethod.java]: s/Mailbox/Mailboxes

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/JmapRuleDTO.java [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+
+import org.apache.james.jmap.api.filtering.Rule;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+public class JmapRuleDTO {
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/JmapRuleDTO.java]: no equals verifiers on these classes?

Comment (chibenwa) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/JmapRuleDTO.java]: No: I never use it for tests as I have integration test for serialization / deserialisation hence those tests are not even needed as the code they would test do not exist

**Comment by aduprat** [comment]
merged
